docs/Declarative-API.md

```diff
@@ -0,0 +1,104 @@
+Formula is a mix of functional, reactive and declarative programming. One aspect that might seem quite unusual is the
+way that it handles asynchronous actions such as RxJava observables, Kotlin Flows, etc. Most developers are used to
+explicitly managing subscription lifecycle.
+```kotlin
+val fetchUserObservable = repository.fetchUser()
+disposables += fetchUserObservable.subscribe { userResult ->
+    // Do something
+}
+```
+
+Formula does things a bit differently. It manages the lifecycle of the asynchronous actions for you. Instead of manually subscribing and unsubscribing,
+you define the conditions for which the asynchronous action should run and the callback which handles events produced by the action.
+```kotlin
+val fetchUserStream = RxStream.fromObservable { repository.fetchUser() }
+fetchUserStream.onEvent { userResult ->
+  // Do something
+}
+```
+
+The logic looks very similar to the first option, but the key difference here is that `Observable.subscribe` hasn't run yet - the execution
+is deferred. It might not be clear why this is useful just from this example, but deferring execution allows us to provide a declarative API.
+For example, we can add conditional logic to only fetch user when user id is set.
+```kotlin
+if (state.userId != null) {
+    val fetchUserStream = RxStream.fromObservable { repository.fetchUser(state.userId) }
+    fetchUserStream.onEvent { userResult ->
+        // Do something with the result
+    }
+}
+```
+
+What if we want to fetch user information only after user clicks on some button to enable this. We can just expand on the conditional logic.
+```kotlin
+if (state.isUserFetchEnabled && state.userId != null) {
+    // Logic here is the same as the previous example
+}
+```
+
+Here, we don’t care what controls `isUserFetchEnabled` boolean. Formula will start execution when these
+conditions are met and will dispose of the action if `state.isUserFetchedEnabled` becomes `false` again.
+
+What if for some unusual reason the userId could change and we would want to refetch? We can
```

Comment:
We might want to change this a little to clarify that by default even though the evaluate function can be run may times, our stream will stay the same and only be subscribed when it is first added to the evaluate output. When I first encountered formula code that did something like RxStream.fromObservable { Observable.timer(...) }, I wasn't immediately sure that formula would make sure my timer wouldn't get restarted each time the evaluate function ran.

Maybe something along these lines to make everything explicit...  
By default because the `evaluate` function can be called many times, the `Stream` is only subscribed to when it is first added to the `evaluate` output. We can modify this behavior by using the `key` param. In this example we refetch the stream any time the userId changes. Without using the key param, even though we pass the `state.userId` into our function, it will only ever run the fetch for the userId that was part of the state when the stream was first added to the `evaluate` output.
```kotlin
if (state.isUserFetchEnabled && state.userId != null) {
    val fetchUserStream = RxStream.fromObservable(key = state.userId) { repository.fetchUser(state.userId) }
    fetchUserStream.onEvent { userResult ->
        // Do something with the result
    }
}
```

Reply:
I also think that because `Compose` has a declarative api we might be able to find some parallels like `LaunchedEffect`, `DisposableEffect` or the `remember(...) { ... }` function and make note of them. That way people who are already familiar with Composes mental model will have a head start in understanding Formulas.